### ISSUE TITLE
lasagne.updates.adam yields float64 updates with float32 0d-array input

### ISSUE THREAD

**Issue by craffel** [Lasagne/Lasagne#402]
Just tracked down this strange behavior.

```Python
>>> import theano
>>> import lasagne
>>> import numpy as np
>>> import theano.tensor as T
>>> x = theano.shared(np.ones((10,), dtype=theano.config.floatX))
>>> f = T.sum(x**2)
>>> u = lasagne.updates.adam(f, [x], beta2=.99)
>>> print [v.type for v in u.values()]
[TensorType(float32, vector),
 TensorType(float32, vector),
 TensorType(float32, vector),
 TensorType(float32, scalar)]
>>> u = lasagne.updates.adam(
...     f, [x], beta2=np.array(.99, dtype=theano.config.floatX))
>>> print [v.type for v in u.values()]
[TensorType(float32, vector),
 TensorType(float64, vector),
 TensorType(float64, vector),
 TensorType(float32, scalar)]
```

If the resulting updates are passed into `theano.function`, a type error is raised because the update for a `float32` values is `float64`.  In this particular case, what's happening is that in `lasagne.adams.updates`, the following:
```
v_t = beta2*v_prev + (1-beta2)*g_t**2
```
`(1 - beta2)` will be given a dtype `float64`:
```Python
>>> (1 - np.array(0., dtype='float32')).dtype
dtype('float64')
```

So, given this, I think we have three options:

1. Implicitly cast things using `utils.floatX` everywhere, which would make the line in question become
`beta2*v_prev + (utils.floatX(1.)-beta2)*g_t**utils.floatX(2.)` (puke)

1. Convert arguments to updates functions to `float`, because if a float is passed instead of a 0d-array with the correct dtype, everything works out fine thanks to Theano's willingness to cast Python scalars to the correct type

1. Tell users (like me) to RTFM, i.e. `beta2 : float` is specified in the docstring, so if I pass anything else (like a numpy scalar), I'm going to have a bad time.

**Comment by benanne** [comment]
I think 3 is out of the question considering that in most situations that substitution works just fine, so I think it makes sense for users to assume it will work here as well (as you did).

2 sounds the least 'invasive' (we just need to add a line or two), and we can add a comment as well explaining the motivation.

1 is ugly but it's not the end of the world. Still, I think I prefer 2.

**Comment by craffel** [comment]
> 2 sounds the least 'invasive' (we just need to add a line or two), and we can add a comment as well explaining the motivation.

Yes, it also raises a broader "parameter validation" question, e.g. to what extent should we verify that the supplied parameters to functions are the correct type, try to cast them when they aren't, and then throw helpful errors when they can't be cast?  That ends up being more than a line or two though, something like:

```Python
if not isinstance(beta2, float):
    try:
        beta2 = float(beta2)
    except:
        raise ValueError('beta2 must be a float, but {} was supplied'.format(beta2))
```
Of course, we could omit the first line as well as the try/except, depending on whether we are OK with people figuring out what the exception caused by `beta2 = float(beta2)` means by themselves.

**Comment by f0k** [comment]
Good catch!

1. Is the correct solution, but you should only cast the `1.`, not the `2`. Exponentiation with an int does not upcast the result.
2. Won't work, because `beta2` might be symbolical. (I make use of that for learning rate or momentum schedules.)
3. Is bad, I agree.

So please go for the first solution! It's common to do something like `one = np.array(1, dtype=theano.config.floatX)` before to make it a little less ugly. If you file a PR, it would be cool if you could check the other update rules for potential problems as well!

**Comment by craffel** [comment]
> Exponentiation with an int does not upcast the result.

Good point!

> Won't work, because beta2 might be symbolical. (I make use of that for learning rate or momentum schedules.)

Very good point!  The docstring should be changed. https://github.com/Lasagne/Lasagne/blob/master/lasagne/updates.py#L538

> So please go for the first solution! It's common to do something like one = np.array(1, dtype=theano.config.floatX)

In this particular case, I think it would be best to do `np.array(1, dtype=value.dtype)`.

> If you file a PR, it would be cool if you could check the other update rules for potential problems as well!

It'll be a bit before I get around to it, I am in deadline mode (again).  It would be cool to have a suite of checks which pass python int, python float, numpy float64 scalar, numpy float32 scalar, and numpy int64 scalar as the different hyperparameters of each update rule and check that the resulting updates are all the correct type.

**Comment by benanne** [comment]
> It'll be a bit before I get around to it, I am in deadline mode (again). It would be cool to have a suite of checks which pass python int, python float, numpy float64 scalar, numpy float32 scalar, and numpy int64 scalar as the different hyperparameters of each update rule and check that the resulting updates are all the correct type.

This sounds like an excellent idea :)

**Comment by f0k** [comment]
> The docstring should be changed.

Yes, we'd need to decide on how to call what type and then fix all the docstrings in all the files. It's unclear whether `float` actually always means a Python float, or whether it includes a numpy float32, or even a Theano scalar. We could invent something like `float-like` (similar to numpy's `array-like`) for things that can be either of those three, or list them separately... that's something to be addressed in a separate issue.

> It'll be a bit before I get around to it, I am in deadline mode (again).

Okay, I can try as well, but I'm in conference mode (greetings from Nice) and I'd like to tackle #11 first.

**Comment by craffel** [comment]
Related: Some `numpy` functions, for a float-type input, return objects of type `numpy.float64`, which also trips Theano up:
```
>>> type(np.tanh(.9))
numpy.float64
```

**Comment by craffel** [comment]
@f0k @benanne can one of you assign this to me so I don't forget to do it at some point?

**Comment by benanne** [comment]
done :)

**Comment by hmendozap** [comment]
I just encountered this problem, as described in the issue and I know that you have been busy and it hasn't been added to the master branch. Three things that I've noticed though and maybe it can help:

1. I also get the same error in sgd, momentum, adagrad and adadelta updates.

2. Theano behaves nicely with python types (ie. float, int) but not so much on numpy types

3. It happens with every kind of update input that can be upcasted eg. my problem was the dropout_output value that was a float32 and that made the gradient to be upcasted to float64. And I have no idea why.

**Comment by sarin1991** [comment]
@craffel @benanne @f0k 
I also faced the same issue while I was running the update steps using adam. If everyone else is busy, I have some free time to look into this now.
Alternatively I was thinking we could also just cast the update variables to the correct type. For e.g.:
`value = param.get_value(borrow=True)
 updates[m_prev] = T.cast(m_t, dtype=value.dtype)
 updates[v_prev] = T.cast(v_t, dtype=value.dtype)
 updates[param] = param - T.cast(step, dtype=value.dtype )`
@hmendozap - I can also take a look at the other updates to see if the same problem is happening

**Comment by nouiz** [comment]
You use use Theano flag warn_floag64=pdb.

Mixing int with float32 frequently will cause upcast to float64. On one
line I saw

1 - something

try

1. - something

**Comment by f0k** [comment]
> If everyone else is busy, I have some free time to look into this now.

This would be welcome! The preferred solution is this one: https://github.com/Lasagne/Lasagne/issues/402#issuecomment-137001684
Although if @nouiz is correct, `1.` would also work, this would be worth investigating.

> I can also take a look at the other updates to see if the same problem is happening

Again, this would be welcome! The DropoutLayer also has the same problem.

**Comment by f0k** [comment]
> Although if @nouiz is correct, `1.` would also work, this would be worth investigating.

No, doesn't make a difference:
```
>>> (1 - np.float32(.5)).dtype
dtype('float64')
>>> (1. - np.float32(.5)).dtype
dtype('float64')
```
We really need to cast the 1 to float32.
```
>>> one = lasagne.utils.floatX(1)
>>> (one - np.float32(.5)).dtype
dtype('float32')
```

**Comment by nouiz** [comment]
I was thinking that in  "1 - n" n was a Theano variable. Theano would
downcast the 1 automatically, but numpy consider all python float as
float64  and all python int as int64 I think.

**Comment by sarin1991** [comment]
Yes, that's exactly what is happening. 

Actually we could just use the as_theano_expression function in utils. It works for both numbers and theano expressions and it handles the casting correctly.
```
>>>import numpy as np
>>>import lasagne
>>>beta2=np.array(.99, dtype='float32')
>>>(1-beta2).dtype
dtype('float64')
>>>(1-lasagne.utils.as_theano_expression(beta2)).dtype
'float32'
```

**Comment by hmendozap** [comment]
@sarin1991 Thanks! I was supposed to look into that but with the thesis I just forgot.

Wouldn't be easier to just cast the whole update dict? Also ````T.cast``` don't think it'll work

**Comment by f0k** [comment]
> Actually we could just use the as_theano_expression function in utils.

Nice idea. The only disadvantage would be that `(1 - as_theano_expression(beta2))` is a Theano expression even if beta2 is just a Python float, while `(floatX(1) - beta2)` becomes a Theano expression only if beta2 is a Theano expression. I.e., if beta2 is a Python float, `floatX(1) - beta2` will be computed while constructing the graph, and `(1 - as_theano_expression(beta2))` will become a graph that is only optimized via constant folding on compilation.
I'm undecided... it seems a bit hackish to replace some arguments by Theano expressions just to get Theano's upcasting rules instead of numpy's, but `one = floatX(1)` is not very elegant either. I think I'd be fine with both.

> Wouldn't be easier to just cast the whole update dict?

Easier to write maybe, but then the computation would take place in float64 (i.e., on the CPU!) and cast to float32 afterwards (i.e., copied back to GPU).

**Comment by Seanny123** [comment]
@hmendozap I'm having the same problem with Adagrad. Is there any sort of workaround possible? Did you find out what the cause of the error was? It doesn't appear to be the type of any of the passed-in parameters.

**Comment by sarin1991** [comment]
The utils.floatX solution actually doesn't work. For instance if the input arguments are python floats then the expression would be upcasted to float64 when it is operated with a python float type
```python
>>>one = np.float32(1)
>>>beta1 = 0.99
>>>(one-beta1).dtype
dtype('float64')
```
So instead I just used utils.as_theano_expression to convert the input arguments

**Comment by hmendozap** [comment]
I'm using [this hack](https://github.com/hmendozap/Lasagne/commit/97ce31c8c8a94640255ad1c2cf119dd2e1118d50)  that I cooked very fast. Please ignore the prints.

@Seanny123 Are you receiving the (hyper)parameters of Adagrad as float32? Otherwise I get this problem

@sarin1991 Are you setting the THEANO_FLAGS correctly?

**Comment by sarin1991** [comment]
@hmendozap that only works when your input arguments are also float32, if you give a python float as input then it won't work. See my previous comment. I haven't got a chance to look at the Dropout Layer yet but I am guessing the fix should be similar
